src/main/java/org/cbioportal/service/impl/StudyViewColumnarServiceImpl.java

```diff
@@ -157,7 +168,80 @@ private StudyViewFilterContext createContext(StudyViewFilter studyViewFilter) {
         return new StudyViewFilterContext(studyViewFilter, customSampleIdentifiers);
     }
 
+    public static List<ClinicalDataCountItem> calculateMissingNaCountsForClinicalDataCountItems(
+        List<ClinicalDataCountItem> clinicalDataCountItems,
+        List<String> filteredAttributes,
+        Map<String, ClinicalDataType> clinicalAttributeDatatypeMap,
+        int filteredSamplesCount,
+        int filteredPatientsCount
+    ) {
+        // Postprocess clinical data count items to adjust NA counts
+        List<ClinicalDataCountItem> combinedClinicalDataCountItems = new ArrayList<>();
+
+        Map<String, ClinicalDataCountItem> clinicalDataCountItemMap = clinicalDataCountItems
+            .stream()
+            .collect(Collectors.toMap(
+                ClinicalDataCountItem::getAttributeId,
+                item -> item
+            ));
+
+        // go over all filtered attributes, not just attributes found in clinicalDataCountItems
+        for (String attributeId: filteredAttributes) {
+            ClinicalDataCountItem clinicalDataCountItem = clinicalDataCountItemMap.get(attributeId);
+            boolean isItemMissing = false;
+
+            if (clinicalDataCountItem == null) {
+                isItemMissing = true;
+                clinicalDataCountItem = new ClinicalDataCountItem();
+                clinicalDataCountItem.setAttributeId(attributeId);
+                clinicalDataCountItem.setCounts(new ArrayList<>());
+            }
+
+            Integer totalClinicalDataCount = clinicalDataCountItem
+                .getCounts()
+                .stream()
+                .map(ClinicalDataCount::getCount)
+                .reduce(0, Integer::sum);
+            // depending on clinical data type we either use filtered sample count or filtered patient count
+            int filteredCount = clinicalAttributeDatatypeMap.get(clinicalDataCountItem.getAttributeId()) == ClinicalDataType.SAMPLE ?
+                filteredSamplesCount: filteredPatientsCount;
+            int casesWithoutClinicalData = filteredCount - totalClinicalDataCount;
 
+            if (casesWithoutClinicalData > 0) {
+                // some of these attributes may be completely missing in clinicalDataCountItem
+                // in case the only attribute value is NA.
+                // we need to manually add those missing items to make sure we have NA counts.
+                if (isItemMissing) {
+                    combinedClinicalDataCountItems.add(clinicalDataCountItem);
+                }
+
+                // find "NA"
+                ClinicalDataCount naClinicalDataCount = clinicalDataCountItem
+                    .getCounts()
+                    .stream()
+                    .filter(c -> c.getValue().equals("NA"))
+                    .findFirst()
+                    .orElse(null);
```

Comment:
really could keep this as an optional and not return null.

then unpack the optional below.
opt.ifPresentOrElse

Reply:
good point!

actually found a better solution

```Java
.orElseGet(() -> {
    // this should only happen when there are multiple studies
    ClinicalDataCount count = new ClinicalDataCount();
    count.setAttributeId(attributeId);
    count.setValue("NA");
    count.setCount(0);
    return count;
});
```